File: src/main/java/ml/duncte123/skybot/commands/music/SkipCommand.java
/*
 * Skybot, a multipurpose discord bot
 *      Copyright (C) 2017  Duncan "duncte123" Sterken
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package ml.duncte123.skybot.commands.music;

import ml.duncte123.skybot.audio.GuildMusicManager;
import ml.duncte123.skybot.audio.TrackScheduler;
import ml.duncte123.skybot.objects.command.MusicCommand;
import ml.duncte123.skybot.utils.Settings;
import net.dv8tion.jda.core.entities.Guild;
import net.dv8tion.jda.core.events.message.guild.GuildMessageReceivedEvent;
import net.dv8tion.jda.core.managers.AudioManager;

public class SkipCommand extends MusicCommand {

    /**
     * This is the executeCommand of the command, the thing you want the command to to needs to be in here
     * @param invoke
     * @param args The command agruments
     * @param event a instance of {@link GuildMessageReceivedEvent GuildMessageReceivedEvent}
     */
    @Override
    public void executeCommand(String invoke, String[] args, GuildMessageReceivedEvent event) {
        Guild guild = event.getGuild();
        GuildMusicManager musicManager = getMusicManager(guild);
        AudioManager audioManager = getAudioManager(guild);
        TrackScheduler scheduler = musicManager.scheduler;

        if(!audioManager.isConnected()){
            sendMsg(event, "I'm not in a voice channel, use `"+ Settings.prefix+"join` to make me join a channel");
            return;
        }

        if(!audioManager.getConnectedChannel().getMembers().contains(event.getMember())){
            sendMsg(event, "I'm sorry, but you have to be in the same channel as me to use any music related commands");
            return;
        }

        if(musicManager.player.getPlayingTrack() == null){
            sendMsg(event, "The player is not playing.");
            return;
        }

        scheduler.nextTrack();

        sendMsg(event,"The current track was skipped.");
    }

    /**
     * The usage instructions of the command
     * @return a String
     */
    @Override
    public String help() {
        // TODO Auto-generated method stub
        return "skips the current track";
    }

    @Override
    public String getName() {
        return "skip";
    }

}

File: src/main/kotlin/ml/duncte123/skybot/commands/music/SkipCommand.kt
/*
 * Skybot, a multipurpose discord bot
 *      Copyright (C) 2017 - 2020  Duncan "duncte123" Sterken & Ramid "ramidzkh" Khan & Maurice R S "Sanduhr32"
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package ml.duncte123.skybot.commands.music

import me.duncte123.botcommons.messaging.MessageUtils.sendMsg
import ml.duncte123.skybot.Author
import ml.duncte123.skybot.Settings.NO_STATIC
import ml.duncte123.skybot.Settings.YES_STATIC
import ml.duncte123.skybot.audio.GuildMusicManager
import ml.duncte123.skybot.objects.TrackUserData
import ml.duncte123.skybot.objects.command.CommandContext
import ml.duncte123.skybot.objects.command.MusicCommand
import kotlin.math.ceil

@Author(nickname = "Sanduhr32", author = "Maurice R S")
class SkipCommand : MusicCommand() {

    init {
        this.name = "skip"
        this.aliases = arrayOf("next", "nexttrack", "skiptrack")
        this.help = "Skips the current track"
    }

    override fun run(ctx: CommandContext) {
        val mng = ctx.audioUtils.getMusicManager(ctx.guild)
        val player = mng.player

        if (player.playingTrack == null) {
            sendMsg(ctx, "The player is not playing.")
            return
        }

        if (!player.playingTrack.isSeekable) {
            sendMsg(ctx, "This track is not seekable")
            return
        }

        val author = ctx.author
        val trackData = player.playingTrack.getUserData(TrackUserData::class.java)

        if (trackData.requester == author.idLong) {
            doSkip(ctx, mng)
            return
        }

        val listeners = getLavalinkManager()
            .getConnectedChannel(ctx.guild)
            .members.filter {
                !it.user.isBot && !(it.voiceState?.isDeafened ?: false)
            }.count()

        val votes = trackData.votes

        var msg = if (votes.contains(author.idLong)) {
            "$NO_STATIC You already voted to skip this song `["
        } else {
            votes.add(author.idLong)
            "$YES_STATIC Successfully voted to skip the song `["
        }

        val skippers = getLavalinkManager().getConnectedChannel(ctx.guild)
            .members.filter { votes.contains(it.idLong) }.count()
        val required = ceil(listeners * .55).toInt()

        msg += "$skippers votes, $required/$listeners needed]`"

        sendMsg(ctx, msg)

        if (skippers >= required) {
            doSkip(ctx, mng)
        }
    }

    private fun doSkip(ctx: CommandContext, mng: GuildMusicManager) {
        val player = mng.player

//        player.seekTo(player.playingTrack.duration)

        mng.scheduler.specialSkipCase()

        if (player.playingTrack == null) {
            sendMsg(
                ctx,
                "Successfully skipped the track.\n" +
                    "Queue is now empty."
            )
        }
    }
}
